tests/unit/ViolationUtilsTest.js

```diff
@@ -0,0 +1,505 @@
+                    {
+                        name: 'missingCategory',
+                        type: 'violation',
+                        userMessage: '',
+                    },
+                    ...transactionViolations,
+                ]),
+            );
+        });
+        it('should handle null category', () => {
+            // setup test conditions
+            transaction.category = null;
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(
+                expect.arrayContaining([
+                    {
+                        name: 'missingCategory',
+                        type: 'violation',
+                        userMessage: '',
+                    },
+                    ...transactionViolations,
+                ]),
+            );
+        });
+        it('should handle undefined category', () => {
+            // setup test conditions
+            transaction.category = undefined;
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(
+                expect.arrayContaining([
+                    {
+                        name: 'missingCategory',
+                        type: 'violation',
+                        userMessage: '',
+                    },
+                    ...transactionViolations,
+                ]),
+            );
+        });
+    });
+
+    describe('policyRequiresTags', () => {
+        beforeEach(() => {
+            policyRequiresTags = true;
+            policyTags = {Lunch: {enabled: true}};
+        });
+
+        it('should handle policyRequiresTags correctly', () => {
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(transactionViolations);
+        });
+
+        it('should not add a missingTag violation when tags are not required', () => {
+            policyRequiresTags = false;
+            transaction.category = 'Drinks'; // or any invalid type
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(expect.arrayContaining(transactionViolations));
+        });
+
+        it('should add a missingTag violation if none is provided and policy requires tags', () => {
+            // setup test conditions
+            transaction.tag = null;
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(
+                expect.arrayContaining([
+                    {
+                        name: 'missingTag',
+                        type: 'violation',
+                        userMessage: '',
+                    },
+                ]),
+            );
+        });
+
+        it('should add a tagOutOfPolicy violation when tag does not exist and policy requires tags', () => {
```

Comment:
This logic is wrong. If the tag is missing, only `missingTag` should be in the returned violations array, and not `tagOutOfPolicy`. Same applies for category

Same for the test below

Reply:
Sorry this was poorly worded -- it was to test if the tag did not exist in the policy (i.e. was not an approved tag), I've rephrased it.